C error when leaving off the parentheses on calling a generic function.

### Describe the bug

```v
struct Arr[T] {
mut:
    data []T
}

fn new_arr[T]() Arr[T] {
    return Arr[T] { }
}

fn main() {
    arr := new_arr[int] // I forgot the ()
    println('Hello')
}
```

### Expected Behavior

Some user friendly error message

### Current Behavior

```
$ v -g run bug.v
bug.v:11:5: warning: unused variable: `arr`
    9 | 
   10 | fn main() {
   11 |     arr := new_arr[int] // I forgot the ()
      |     ~~~
   12 |     println('Hello')
   13 | }
/tmp/v_1000/bug.17236921576658051164.tmp.c:932: error: identifier expected
builder error: 
==================
...
$
```

### Reproduction Steps

Code attached above

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

0.3.4 ac32d2a

### Environment details (OS name and version, etc.)

V full version: V 0.3.4 ac32d2a

vexe mtime: 2023-06-16 15:40:56


CC version: cc (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

--- a/vlib/v/checker/assign.v
+++ b/vlib/v/checker/assign.v
@@ -514,7 +514,7 @@ fn (mut c Checker) assign_stmt(mut node ast.AssignStmt) {
 					right.pos())
 			} else if (!right_sym.info.is_anon && return_sym.kind == .any)
 				|| (return_sym.info is ast.Struct && (return_sym.info as ast.Struct).is_generic) {
-				panic('cannot assign `${right}` as a generic function variable')
+				c.error('cannot assign `${right}` as a generic function variable', right.pos())
 			}
 		}
 		if left_type.is_any_kind_of_pointer() && !left.is_auto_deref_var() {

--- a/vlib/v/checker/assign.v
+++ b/vlib/v/checker/assign.v
@@ -507,6 +507,15 @@ fn (mut c Checker) assign_stmt(mut node ast.AssignStmt) {
 			c.error('cannot copy map: call `move` or `clone` method (or use a reference)',
 				right.pos())
 		}
+		if left_sym.kind == .function && right_sym.info is ast.FnType {
+			return_sym := c.table.sym(right_sym.info.func.return_type)
+			if return_sym.kind == .placeholder {
+				c.error('cannot assign function `${right_sym.name}` to variable', right.pos())
+			} else if (!right_sym.info.is_anon && return_sym.kind == .any)
+				|| (return_sym.info is ast.Struct && (return_sym.info as ast.Struct).is_generic) {
+				panic('cannot assign function `${right_sym.name}` with generic return type `${return_sym.name}` to a variable')
+			}
+		}
 		if left_type.is_any_kind_of_pointer() && !left.is_auto_deref_var() {
 			if !c.inside_unsafe && node.op !in [.assign, .decl_assign] {
 				// ptr op=

--- a/vlib/v/checker/assign.v
+++ b/vlib/v/checker/assign.v
@@ -510,10 +510,11 @@ fn (mut c Checker) assign_stmt(mut node ast.AssignStmt) {
 		if left_sym.kind == .function && right_sym.info is ast.FnType {
 			return_sym := c.table.sym(right_sym.info.func.return_type)
 			if return_sym.kind == .placeholder {
-				c.error('cannot assign function `${right_sym.name}` to variable', right.pos())
+				c.error('unkown return type: cannot assign `${right}` as a function variable',
+					right.pos())
 			} else if (!right_sym.info.is_anon && return_sym.kind == .any)
 				|| (return_sym.info is ast.Struct && (return_sym.info as ast.Struct).is_generic) {
-				panic('cannot assign function `${right_sym.name}` with generic return type `${return_sym.name}` to a variable')
+				panic('cannot assign `${right}` as a generic function variable')
 			}
 		}
 		if left_type.is_any_kind_of_pointer() && !left.is_auto_deref_var() {

--- a/vlib/v/checker/tests/generic_fn_infinite_loop_limit_err.out
+++ b/vlib/v/checker/tests/generic_fn_infinite_loop_limit_err.out
@@ -5,6 +5,13 @@ vlib/v/checker/tests/generic_fn_infinite_loop_limit_err.vv:6:2: warning: unused 
       |     ~~~
     7 |     // ff1 := f1[int] <-- is a valid usage with generic return types that are not generic structs
     8 | }
+vlib/v/checker/tests/generic_fn_infinite_loop_limit_err.vv:6:9: error: cannot assign `f1` as a generic function variable
+    4 | 
+    5 | fn main() {
+    6 |     ff1 := f1 // <-- missing e.g. `[int]`
+      |            ~~
+    7 |     // ff1 := f1[int] <-- is a valid usage with generic return types that are not generic structs
+    8 | }
 vlib/v/checker/tests/generic_fn_infinite_loop_limit_err.vv:1:1: error: generic function visited more than 10000 times
     1 | fn f1[T](x T, i int) T {
       | ~~~~~~~~~~~~~~~~~~~~~~

--- a/vlib/v/checker/tests/unknown_array_fn_type_in_struct_field.out
+++ b/vlib/v/checker/tests/unknown_array_fn_type_in_struct_field.out
@@ -5,3 +5,10 @@ Did you mean `[2]fn (u32) UnknownThing`?
       |                         ~~~~~~~~~~~~
     3 | }
     4 |
+vlib/v/checker/tests/unknown_array_fn_type_in_struct_field.vv:6:18: error: cannot assign function `fn (u32) UnknownThing` to variable
+    4 | 
+    5 | fn (virt Virt) caller() {
+    6 |     func := virt.fns[0]
+      |                     ~~~
+    7 |     func(5)
+    8 | }

--- a/vlib/v/checker/tests/unknown_array_fn_type_in_struct_field.out
+++ b/vlib/v/checker/tests/unknown_array_fn_type_in_struct_field.out
@@ -5,7 +5,7 @@ Did you mean `[2]fn (u32) UnknownThing`?
       |                         ~~~~~~~~~~~~
     3 | }
     4 |
-vlib/v/checker/tests/unknown_array_fn_type_in_struct_field.vv:6:18: error: cannot assign function `fn (u32) UnknownThing` to variable
+vlib/v/checker/tests/unknown_array_fn_type_in_struct_field.vv:6:18: error: unkown return type: cannot assign `virt.fns[0]` as a function variable
     4 | 
     5 | fn (virt Virt) caller() {
     6 |     func := virt.fns[0]

